packages/hash/shared/src/ProsemirrorSchemaManager.ts

```diff
@@ -391,39 +391,239 @@ export class ProsemirrorSchemaManager {
+   * @todo need to support variants here
    */
   async replaceNodeWithRemoteBlock(
     draftBlockId: string,
     targetComponentId: string,
+    targetVariant: BlockVariant,
     node: ProsemirrorNode<Schema>,
-    getPos: () => number,
+    pos: number,
   ) {
     const { view } = this;
 
+    // @todo consider separating the logic that uses a view to something else
     if (!view) {
       throw new Error("Cannot trigger replaceNodeWithRemoteBlock without view");
     }
 
-    const entityStoreState = entityStorePluginState(view.state);
-    const newNode = await this.createRemoteBlock(
+    const [tr, newNode] = await this.createRemoteBlockTr(
       targetComponentId,
-      entityStoreState.store,
       draftBlockId,
+      targetVariant,
     );
 
-    /**
-     * The code below used to ensure the cursor was positioned
-     * within the new node, depending on its type, but because we
-     * now want to trigger saves when we change node type, and
-     * because triggering saves can mess up the cursor position,
-     * we're currently not re-focusing the editor view.
-     */
-
-    const pos = getPos();
-    const { tr } = view.state;
-
     tr.replaceRangeWith(pos, pos + node.nodeSize, newNode);
     view.dispatch(tr);
   }
+
+  // @todo handle empty variant properties
+  // @todo handle saving the results of this
+  // @todo handle non-intermediary entities
+  async createRemoteBlockTr(
+    targetComponentId: string,
+    draftBlockId: string | null,
+    targetVariant: BlockVariant,
+  ) {
+    if (!this.view) {
+      throw new Error("Cannot trigger createRemoteBlockTr without view");
+    }
+
+    const { tr } = this.view.state;
+    const meta = await this.fetchAndDefineBlock(targetComponentId);
+
+    let blockIdForNode = draftBlockId;
+
+    if (blockIdForNode) {
+      const entityStoreState = entityStorePluginState(this.view.state);
+
+      const blockEntity = entityStoreState.store.draft[blockIdForNode];
+
+      if (!blockEntity || !isDraftBlockEntity(blockEntity)) {
+        throw new Error("draft id does not belong to a block");
+      }
+
+      if (targetComponentId === blockEntity.properties.componentId) {
+        /**
+         * I've temporarily made it so all changes involved text result in
+         * creating a new variant entity, instead of updating the one that's
+         * there. That's because the save mechanism doesn't yet know to process
+         * updates to entities that aren't in the prosemirror document. This
+         * forces the update to be saved anyway, at cost of throw away variant
+         * entities.
+         *
+         * @todo update the existing entity where possible, instead of
+         * creating new ones
+         */
+        if (
+          !blockComponentRequiresText(meta.componentSchema)
+          // ||
+          // // isTextContainingEntityProperties(
+          // //   blockEntity.properties.entity.properties,
+          // // )
```

Comment:
Do we need these commented out lines? If we are planning to enable them, let's add a `@todo`. Otherwise we can just get rid of them I guess.